src/appleseed/renderer/modeling/phasefunction/henyeyphasefunction.cpp

```diff
@@ -0,0 +1,323 @@
+//
+// Copyright (c) 2017 Artem Bishev, The appleseedhq Organization
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+// Interface header.
+#include "henyeyphasefunction.h"
+
+// appleseed.foundation headers.
+#include "foundation/math/basis.h"
+#include "foundation/math/sampling/mappings.h"
+#include "foundation/utility/api/specializedapiarrays.h"
+#include "foundation/utility/containers/dictionary.h"
+
+// appleseed.renderer headers.
+#include "renderer/kernel/shading/shadingray.h"
+#include "renderer/global/globaltypes.h"
+#include "renderer/modeling/input/inputarray.h"
+#include "renderer/modeling/phasefunction/phasefunction.h"
+
+
+namespace renderer
+{
+
+namespace
+{
+    const char* Model = "henyey_phasefunction";
+}
+
+//
+// Henyey-Greenstein phase function.
+//
+// https://www.astro.umd.edu/~jph/HG_note.pdf
+//
+
+class HenyeyPhaseFunction
+    : public PhaseFunction
+{
+  public:
+    HenyeyPhaseFunction(
+        const char*             name,
+        const ParamArray&       params)
+        : PhaseFunction(name, params)
+    {
+        m_inputs.declare("scattering", InputFormatSpectralReflectance);
+        m_inputs.declare("scattering_multiplier", InputFormatFloat, "1.0");
+        m_inputs.declare("extinction", InputFormatSpectralReflectance);
+        m_inputs.declare("extinction_multiplier", InputFormatFloat, "1.0");
+        m_inputs.declare("average_cosine", InputFormatFloat, "1.0");
+    }
+
+    virtual void release() APPLESEED_OVERRIDE
+    {
+        delete this;
+    }
+
+    virtual const char* get_model() const APPLESEED_OVERRIDE
+    {
+        return Model;
+    }
+        
+    virtual bool is_homogeneous() const
+    {
+        return true;
+    }
+
+    virtual void prepare_inputs(
+        foundation::Arena&          arena,
+        const ShadingRay&           shading_point,
+        void*                       data
+        ) const APPLESEED_OVERRIDE
```

Comment:
Move closing parenthesis to previous line:

    virtual void prepare_inputs(
        foundation::Arena&          arena,
        const ShadingRay&           shading_point,
        void*                       data) const APPLESEED_OVERRIDE